test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -119,119 +126,98 @@ func BenchmarkPerfScheduling(b *testing.B) {
-	}
-
 	finalFunc, podInformer, clientset := mustSetupScheduler()
 	defer finalFunc()
 
-	var nodePreparer testutils.TestNodePreparer
-	if nodeSpec != nil {
-		nodePreparer = framework.NewIntegrationTestNodePreparerWithNodeSpec(
-			clientset,
-			[]testutils.CountToStrategy{{Count: test.Nodes.Num, Strategy: nodeStrategy}},
-			nodeSpec,
-		)
-	} else {
-		nodePreparer = framework.NewIntegrationTestNodePreparer(
-			clientset,
-			[]testutils.CountToStrategy{{Count: test.Nodes.Num, Strategy: nodeStrategy}},
-			"scheduler-perf-",
-		)
-	}
-
+	nodePreparer := getNodePreparer(test.Nodes, clientset)
 	if err := nodePreparer.PrepareNodes(); err != nil {
 		klog.Fatalf("%v", err)
 	}
 	defer nodePreparer.CleanupNodes()
 
-	config := testutils.NewTestPodCreatorConfig()
-	config.AddStrategy(setupNamespace, test.InitPods.Num, setupPodStrategy)
-	podCreator := testutils.NewTestPodCreator(clientset, config)
-	podCreator.CreatePods()
-
-	for {
-		scheduled, err := getScheduledPods(podInformer)
-		if err != nil {
-			klog.Fatalf("%v", err)
-		}
-		if len(scheduled) >= test.InitPods.Num {
-			break
-		}
-		klog.Infof("got %d existing pods, required: %d", len(scheduled), test.InitPods.Num)
-		time.Sleep(1 * time.Second)
-	}
-
-	scheduled := int32(0)
-	completedCh := make(chan struct{})
-	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		UpdateFunc: func(old, cur interface{}) {
-			curPod := cur.(*v1.Pod)
-			oldPod := old.(*v1.Pod)
-
-			if len(oldPod.Spec.NodeName) == 0 && len(curPod.Spec.NodeName) > 0 {
-				if atomic.AddInt32(&scheduled, 1) >= int32(test.PodsToSchedule.Num) {
-					completedCh <- struct{}{}
-				}
-			}
-		},
-	})
+	createPods(setupNamespace, test.InitPods, clientset)
+	waitNumPodsScheduled(test.InitPods.Num, podInformer)
 
 	// start benchmark
 	b.ResetTimer()
 
-	// Start measuring throughput
+	// Start test data collectors.
 	stopCh := make(chan struct{})
-	throughputCollector := newThroughputCollector(podInformer)
-	go throughputCollector.run(stopCh)
+	collectors := getTestDataCollectors(test, podInformer)
+	for _, collector := range collectors {
+		go collector.run(stopCh)
+	}
 
-	// Scheduling the main workload
-	config = testutils.NewTestPodCreatorConfig()
-	config.AddStrategy(testNamespace, test.PodsToSchedule.Num, testPodStrategy)
-	podCreator = testutils.NewTestPodCreator(clientset, config)
-	podCreator.CreatePods()
+	// Schedule the main workload
+	createPods(testNamespace, test.PodsToSchedule, clientset)
+	waitNumPodsScheduled(test.InitPods.Num+test.PodsToSchedule.Num, podInformer)
 
-	<-completedCh
 	close(stopCh)
-
 	// Note: without this line we're taking the overhead of defer() into account.
 	b.StopTimer()
 
-	setNameLabel := func(dataItem *DataItem) DataItem {
-		if dataItem.Labels == nil {
-			dataItem.Labels = map[string]string{}
+	var dataItems []DataItem
+	for _, collector := range collectors {
+		dataItems = append(dataItems, collector.collect(b)...)
```

Comment:
I was trying to avoid the case of passing anything additional to collectors. Though as long as `collector` is a private interface, I have no objections.

Reply:
I don't feel strongly about introducing the interface. If you don't like it, we can get rid of it.

Reply:
Nothing against private bits here. Though, I would rather prefer:
```
type testDataCollector interface {
	run(stopCh chan struct{})
	collect() []DataItem
}
```
and have:
```
func newMetricsCollector(config metricsCollectorConfig, labels [map]string) *metricsCollector
func newThroughputCollector(podInformer coreinformers.PodInformer, labels [map]string) *throughputCollector
```
instead to simplify the interface.

Reply:
Sounds good.